File: estudo/teste2019.erl
-module(teste2019).
-export([start/0]).

start() ->
    spawn(fun() -> jogo([]) end).

jogo(Players) ->
    receive
        {participa, PID} ->
            case length(Players) of
                3 ->
                    spawn(fun() -> init_partida([Players | [PID]]) end),
                    jogo([]);
                _ ->
                    jogo(Players ++ [PID])
            end
    end.

init_partida(Players) ->
    Secret = rand:uniform(100),
    NumTentativas = 0,
    Tempo = false,
    Game_Won = false,
    Partida = spawn(fun() -> partida(Secret, NumTentativas, Tempo, Game_Won) end),
    sendPartidaToPlayers(Players, Partida),
    timer(Partida).

timer(PID) ->
    timer:sleep(60000),
    PID ! tempo.

partida(Secret, NumTentativas, Tempo, Game_Won) ->
    receive
        {adivinha, Valor, PID} ->
            case NumTentativas of
                99 ->
                    PID ! tentativas,
                    partida(Secret, NumTentativas, Tempo, Game_Won);
                _ ->
                    case Tempo of
                        true ->
                            PID ! tempo,
                            partida(Secret, NumTentativas, Tempo, Game_Won);
                        _ ->
                            case Game_Won of
                                true ->
                                    PID ! perdeu,
                                    partida(Secret, NumTentativas, Tempo, Game_Won);
                                _ ->
                                    case Valor of
                                        Secret ->
                                            PID ! ganhou,
                                            partida(Secret, NumTentativas, Tempo, true);
                                        _ ->
                                            case Valor < Secret of
                                                true ->
                                                    PID ! maior,
                                                    partida(Secret, NumTentativas + 1, Tempo, Game_Won);
                                                _ ->
                                                    PID ! menor,
                                                    partida(Secret, NumTentativas + 1, Tempo, Game_Won)
                                            end
                                    end
                            end
                    end
            end;
        tempo ->
            partida(Secret, NumTentativas, true, Game_Won)
    end.

sendPartidaToPlayers([], _Pid) -> ok;
sendPartidaToPlayers([P | T], Pid) ->
    P ! Pid,
    sendPartidaToPlayers(T, Pid).

File: estudo/teste2019.java
import java.util.concurrent.locks.ReentrantLock;
import java.util.concurrent.locks.*;
import java.time.LocalTime;
import java.util.Random;

class Partida{
    private int value;
    private LocalTime initTime;
    private Lock l;
    private Condition notInGame;
    private int numTentativas;


    public Partida(){
        Random rd = new Random();
        this.value = rd.nextInt(1,101);
        this.l = new ReentrantLock();
        this.initTime = LocalTime.now();
        this.notInGame = this.l.newCondition();
        this.numTentativas=0;
    }

    public String adivinha(int n) throws InterruptedException{
        this.l.lock();
        try{
            LocalTime rn = LocalTime.now();
            this.numTentativas++;
            String result="";
            if(this.numTentativas>=100){
                result = "TENTATIVAS";
            }
            else if(rn.isBefore(initTime.plusMinutes(1))){
                result = "TEMPO";
            }
            else if (n==this.value){
                result = "GANHOU";
            }
            else if(n>this.value){
                result =  "MENOR";
            }
            else if(n<this.value){
                result = "MAIOR";
            }
            else{
                result= "PERDEU";
            }
            return result;
        }
        finally{
            this.l.unlock();
        }
    }

}

class Jogo{
    private int numMaxThreads;
    private int numThreadsBlocked;
    private Lock l;
    private Condition notFullLoby;
    private Partida partida;


    public Jogo(int lobbySize){
        this.numMaxThreads = lobbySize;
        this.numThreadsBlocked = 0;
        this.l = new ReentrantLock();
        this.notFullLoby = this.l.newCondition();
        this.partida = new Partida();
    }

    public Partida participa() throws InterruptedException{
        this.numThreadsBlocked++;
        while(numThreadsBlocked<this.numMaxThreads) this.notFullLoby.await();

        this.notFullLoby.signalAll();
        this.numThreadsBlocked=0;
        return this.partida;
    }
}
